### PR TITLE
Langchain: Ensure that the Elasticsearch Query Translator functions accurately w…

### PR DESCRIPTION
**Pull request by khabouss**
Description:
Addresses a problem where the Date type within an Elasticsearch SelfQueryRetriever would encounter difficulties in generating a valid query.

Issue: #17042

### CODE REVIEW

**Comment by baskaryan** [comment]
cc @joemcelroy

**Review by baskaryan — COMMENTED** [review]
could we add a unit test for this

**Comment by khabouss** [comment]
@baskaryan, I have included unit tests.

**Review by joemcelroy — APPROVED** [review]
thanks!

File: libs/langchain/langchain/retrievers/self_query/elasticsearch.py [hunk 1]
```diff
@@ -61,11 +61,25 @@ def visit_comparison(self, comparison: Comparison) -> Dict:
...
+                            }
+                        }
+                    }
+            else:
+                return {
+                    "range": {
+                        field: {
+                            self._format_func(comparison.comparator): comparison.value
+                        }
+                    }
```
Comment (maxjakob) [inline libs/langchain/langchain/retrievers/self_query/elasticsearch.py:81]: Small suggestion to improve readability:

```py
value = comparison.value
if isinstance(comparison.value, dict) and "date" in comparison.value:
    value = comparison.value["date"]

return {
    "range": {
        field: {
            self._format_func(comparison.comparator): value
        }
    }
}
```

**Review by maxjakob — COMMENTED** [review]
Thank you so much for your contribution! 🙏  
Just added a small suggestion.

File: libs/langchain/langchain/retrievers/self_query/elasticsearch.py [hunk 2]
```diff
@@ -61,11 +61,12 @@ def visit_comparison(self, comparison: Comparison) -> Dict:
...
 
         if is_range_comparator:
-            return {
-                "range": {
-                    field: {self._format_func(comparison.comparator): comparison.value}
+            value = comparison.value
+            if isinstance(comparison.value, dict) and "date" in comparison.value:
+                value = comparison.value["date"]
+                return {
+                    "range": {field: {self._format_func(comparison.comparator): value}}
```
Comment (maxjakob) [inline libs/langchain/langchain/retrievers/self_query/elasticsearch.py:68]:
```suggestion
            return {
                "range": {field: {self._format_func(comparison.comparator): value}}
```
I think indentation got mixed up here

**Review by baskaryan — APPROVED** [review]

**Pull request merged**